### PR TITLE
feat: created new opentelemetry package with tracing functionality

### PR DESCRIPTION
**Pull request by CyntiBinti**
JIRA[CPREL-930]

This creates a new opentelemetry package, where all tests are passing; and can be --required also if the consuming repo wants to use it in this way

Release-As: 0.1.0

### CODE REVIEW

File: packages/opentelemetry/lib/index.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+} = require('@opentelemetry/exporter-trace-otlp-proto');
+const { NoopSpanProcessor } = require('@opentelemetry/sdk-trace-base');
+const logger = require('@dotcom-reliability-kit/logger');
+
+/**
+ * Set up OpenTelemetry tracing.
+ *
+ * @param {object} options
+ *      Options passed into the function parameter.
+ * @param {string | null} [options.tracesEndpoint]
```
Comment (CyntiBinti) [inline packages/opentelemetry/lib/index.js:22]: This line is oddly failing linting as it's saying:

```
/Users/cynthia.mbulu/Documents/code/dotcom-reliability-kit/packages/opentelemetry/lib/index.js
  22:1  error  Sentences should start with an uppercase character  jsdoc/require-description-complete-sentence
```

but this linting error doesn't make sense here 😭

Comment (rowanmanning) [reply packages/opentelemetry/lib/index.js:22]: Oh it's the JSDoc plugin being a pain. it sees the `.` in `e.g.` as the end of a sentence 🤦 maybe rephrase as:

> The URL to send OpenTelemetry trace segments to, for example http://localhost:4318/v1/traces.

File: .release-please-manifest.json [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
+	"packages/errors": "2.2.0",
+	"packages/eslint-config": "2.0.1",
+	"packages/fetch-error-handler": "0.1.2",
+	"packages/log-error": "3.1.7",
+	"packages/logger": "2.4.2",
+	"packages/middleware-log-errors": "3.0.9",
+	"packages/middleware-render-error-info": "4.1.1",
+	"packages/serialize-error": "2.2.1",
+	"packages/serialize-request": "2.2.1",
+	"packages/opentelemetry": "0.0.0"
```
Comment (rowanmanning) [inline .release-please-manifest.json:13]: **nitpick:** the diff indicates that all lines in this file have changed, can we revert the whitespace so that the blame is maintained?

File: packages/opentelemetry/lib/index.js (hunk 1 above)
Comment (CyntiBinti) [reply packages/opentelemetry/lib/index.js:22]: OMG 😅  - it's so nitpicky. Thank you!!

File: README.md [hunk 3]
```diff
@@ -37,6 +37,9 @@ We maintain documentation in this repo:
     * **[@dotcom-reliability-kit/middleware-render-error-info](./packages/middleware-render-error-info/#readme):**<br/>
       Express middleware to render error information in a browser in a way that makes local debugging easier
 
+    * **[@dotcom-reliability-kit/opentelemetry](./packages/opentelemetry/#readme):**<br/>
+      A utility function using Opentelemtery, providing health, performance, and latency data about an FT system
```
Comment (rowanmanning) [inline README.md:41]: **suggestion:** thinking about a description for this package and I wonder if we can simplify it a bit by leaning on the OpenTelemetry docs?

```suggestion
      An [OpenTelemetry](https://opentelemetry.io/docs/what-is-opentelemetry/) client that's preconfigured for drop-in use in FT apps.
```

File: packages/opentelemetry/package.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
+{
+	"name": "@dotcom-reliability-kit/opentelemetry",
+	"version": "0.0.0",
+	"description": "A utility function using Opentelemtery, providing health, performance, and latency data about an FT system",
```
Comment (rowanmanning) [inline packages/opentelemetry/package.json:4]: **suggestion:** see previous comment about the description:

```suggestion
	"description": "An OpenTelemetry client that's preconfigured for drop-in use in FT apps.",
```

File: packages/opentelemetry/README.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+
+## @dotcom-reliability-kit/opentelemetry
+
+This module is part of [FT.com Reliability Kit](https://github.com/Financial-Times/dotcom-reliability-kit#readme).
```
Comment (rowanmanning) [inline packages/opentelemetry/README.md:4]: **suggestion:** I don't think we need a full README in this PR but we could add the description in as well as a note to say this package is experimental:

```suggestion
An [OpenTelemetry](https://opentelemetry.io/docs/what-is-opentelemetry/) client that's preconfigured for drop-in use in FT apps. This module is part of [FT.com Reliability Kit](https://github.com/Financial-Times/dotcom-reliability-kit#readme).

> [!WARNING]<br />
> This package is in beta and hasn't been tested extensively in production yet. Feel free to use, and any feedback is greatly appreciated.
```

File: packages/opentelemetry/lib/index.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+const {
+	SemanticResourceAttributes
+} = require('@opentelemetry/semantic-conventions');
+const appInfo = require('@dotcom-reliability-kit/app-info');
+const {
+	OTLPTraceExporter
+} = require('@opentelemetry/exporter-trace-otlp-proto');
+const { NoopSpanProcessor } = require('@opentelemetry/sdk-trace-base');
+const logger = require('@dotcom-reliability-kit/logger');
+// OpenTelemetry Traces Endpoint to which to export traces (e.g. http:localhost:4318/v1/traces).
```
Comment (rowanmanning) [inline packages/opentelemetry/lib/index.js:16]: **suggestion:** left in by accident?

```suggestion

```

File: packages/opentelemetry/lib/index.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * Set up OpenTelemetry tracing.
+ *
+ * @param {object} options
+ *      Options passed into the function parameter.
+ * @param {string | null} [options.tracesEndpoint]
+ *      The URL to send OpenTelemetry trace segments to, for example http://localhost:4318/v1/traces.
+ */
+function opentelemetry({ tracesEndpoint }) {
```
Comment (rowanmanning) [inline packages/opentelemetry/lib/index.js:25]: **suggestion:** not a blocker but maybe the entire options object should be optional too? This would match other packages [like crash handler](https://github.com/Financial-Times/dotcom-reliability-kit/blob/b41508441601f51a3dec4d0ad0e0d79a119c6f28/packages/crash-handler/lib/index.js#L21) and it means that if you don't want to specify any options you can call the function without having to pass in an empty object:

```suggestion
/**
 * Set up OpenTelemetry tracing.
 *
 * @param {object} [options]
 *      Options passed into the function parameter.
 * @param {string | null} [options.tracesEndpoint]
 *      The URL to send OpenTelemetry trace segments to, for example http://localhost:4318/v1/traces.
 */
function opentelemetry({ tracesEndpoint } = {}) {
```

Comment (rowanmanning) [inline packages/opentelemetry/lib/index.js:25]: **nitpick:** I think maybe we could make this method more descriptive of what it's doing, maybe something like `setupOpenTelemetry`?

File: packages/opentelemetry/loadOtel.js
Comment (rowanmanning) [inline packages/opentelemetry/loadOtel.js:1]: **nitpick:** I think the `Otel` in this filename is redundant because it'll be referenced as

```
@dotcom-reliability-kit/opentelemetry/loadOtel
```

Maybe something like `setup` describes more what it's doing?

```
@dotcom-reliability-kit/opentelemetry/setup
```

**Review by rowanmanning — CHANGES_REQUESTED** [review]
Nice :) looking generally good aside from tests which I know you're working on! I've made a few suggestions in comments.

File: packages/opentelemetry/package.json [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+		"type": "git",
+		"url": "https://github.com/Financial-Times/dotcom-reliability-kit.git",
+		"directory": "packages/opentelemetry"
+	},
+	"homepage": "https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/opentelemetry#readme",
+	"bugs": "https://github.com/Financial-Times/dotcom-reliability-kit/issues?q=label:\"package: opentelemetry\"",
+	"license": "MIT",
+	"engines": {
+		"node": "16.x || 18.x || 20.x",
+		"npm": "7.x || 8.x || 9.x"
```
[the remark this answers is not in the document]
Comment (CyntiBinti) [reply packages/opentelemetry/package.json:15]: This makes sense - will remove and bear in mind for next time

File: packages/opentelemetry/lib/index.js (hunk 6 above)
Comment (CyntiBinti) [reply packages/opentelemetry/lib/index.js:16]: Yes sorry - Was trying to figure out the JSDoc nitpicky issue  - will remove thanks!

File: packages/opentelemetry/lib/index.js (hunk 7 above)
Comment (CyntiBinti) [reply packages/opentelemetry/lib/index.js:25]: Yeah this flexibility makes sense

Comment (CyntiBinti) [reply packages/opentelemetry/lib/index.js:25]: Yeah good point

File: packages/opentelemetry/loadOtel.js
Comment (CyntiBinti) [reply packages/opentelemetry/loadOtel.js:1]: Much clearer

File: .release-please-manifest.json (hunk 2 above)
Comment (rowanmanning) [reply .release-please-manifest.json:13]: @CyntiBinti the change is that all the two-space indentation has changed to tabs. This would normally be OK but Release Please will switch it back to two spaces so there'll be a lot of weird churn in this file. You can reduce the diff by finding all tabs and replacing them with two spaces.

**Review by rowanmanning — APPROVED** [review]
Looking excellent! Thanks for all your perseverance with the tests – I know it was long but hopefully useful to run through some of the less well-known Jest concepts. No notes: let's get this merged 🚀

**Comment by CyntiBinti** [comment]
> Looking excellent! Thanks for all your perseverance with the tests – I know it was long but hopefully useful to run through some of the less well-known Jest concepts. No notes: let's get this merged 🚀

You've been instrumental in me having a better understanding of Jest - so TYSM ❤️ 🙇🏽‍♀️
